### PR TITLE
gpt-oss

### PR DESCRIPTION
**Pull request by eitanturok**
gpt-oss inference works in tinygrad but is slow!

todo:
1.  ✅ mxfp4 dequantization
2. ✅  sliding window attention
3. ✅ custom head_dim
4. ✅ `MixtureFeedForward` (differs from OLMoE's `MixtureFeedForward`)
5. ✅ yarn
6. ✅ attention sinks
7. ✅ tinygrad matches pytorch with small, dummy model
8. ✅ tinygrad matches pytorch with real model
9. ✅ start measuring stuff (throughput, FLOPs, etc.)
10. ✅  bf16, not fp32
11. ✅ fix jit
12. use gguf model [https://huggingface.co/unsloth/gpt-oss-20b-GGUF/tree/main](https://huggingface.co/unsloth/gpt-oss-20b-GGUF/tree/main)
13. add a bunch of `realize`, `contiguous`
14. beam
15. find slow kernels

### CODE REVIEW

**Comment by wozeparrot** [comment]
we eventually want `apps.llm` to be gpt-oss, but I think its good to have it as another example if it can reuse stuff from llama

**Comment by eitanturok** [comment]
`VIZ=1 DEBUG=2 PYTHONPATH=. python tinygrad/apps/llm2.py` generates only spaces (token_id 220). Why?

This cmd takes up 80.5 GB of memory which seems way to high. Why?

Screenshot 2025-10-26 at 10 37 39 AM

We also generate tons of topk kernels with different dims. Is `topk` called recursively or something?
```sh
*** CUDA    5124 r_32_16_180                                    arg  4 mem  80.45 GB tm     12.90us/  7631.46ms (     15 GFLOPS   15|46     GB/s) ['linear', 'topk']
*** CUDA    5125 E_8_2_2                                        arg  2 mem  80.45 GB tm      7.84us/  7631.47ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5126 r_32_16_2                                      arg  2 mem  80.45 GB tm      8.54us/  7631.47ms (      2 GFLOPS    0|5      GB/s) ['topk']
*** CUDA    5127 E_16_2                                         arg  2 mem  80.45 GB tm      7.94us/  7631.48ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5128 E_4_2_2_2                                      arg  2 mem  80.45 GB tm      8.22us/  7631.49ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5129 E_8_2_2n1                                      arg  2 mem  80.45 GB tm      7.36us/  7631.50ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5130 E_16_2                                         arg  2 mem  80.45 GB tm      7.87us/  7631.51ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5131 E_2_4_2_2                                      arg  2 mem  80.45 GB tm      8.58us/  7631.51ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5132 E_4_4_2                                        arg  2 mem  80.45 GB tm      7.55us/  7631.52ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5133 E_8_2_2n1                                      arg  2 mem  80.45 GB tm      6.98us/  7631.53ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5134 E_16_2                                         arg  2 mem  80.45 GB tm      7.81us/  7631.54ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5135 E_8_2_2n2                                      arg  2 mem  80.45 GB tm      7.84us/  7631.54ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5136 E_2_8_2                                        arg  2 mem  80.45 GB tm      8.22us/  7631.55ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5137 E_4_4_2                                        arg  2 mem  80.45 GB tm      7.14us/  7631.56ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5138 E_8_2_2n1                                      arg  2 mem  80.45 GB tm      7.07us/  7631.57ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5139 E_16_2                                         arg  2 mem  80.45 GB tm      7.30us/  7631.57ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5140 E_16_2n1                                       arg  2 mem  80.45 GB tm      8.48us/  7631.58ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5141 E_2_8_2                                        arg  2 mem  80.45 GB tm      7.36us/  7631.59ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5142 E_4_4_2                                        arg  2 mem  80.45 GB tm      7.58us/  7631.60ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5143 E_8_2_2n1                                      arg  2 mem  80.45 GB tm      7.65us/  7631.60ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5144 E_16_2                                         arg  2 mem  80.45 GB tm      7.68us/  7631.61ms (      0 GFLOPS    0|0      GB/s) ['topk']
*** CUDA    5145 r_4_2_2_2_16_2                                 arg  2 mem  80.45 GB tm      8.00us/  7631.62ms (      3 GFLOPS    0|5      GB/s) ['topk']
```

I think these issues might be because I'm not dequantizing the mpxfp4 weights correctly...

**Comment by eitanturok** [comment]
Inference officially works. The tinygrad implementation matches torch/hf!

Check for yourself:
`PYTHONPATH=. FAKEWEIGHTS=1 python -m unittest test.external.external_test_gpt_oss` tests that gpt-oss with 1-layer and fakeweights matches in tinygrad and pytorch

`PYTHONPATH=. python tinygrad.apps.llm2.py` and `TORCH=1 PYTHONPATH=. python tinygrad.apps.llm2.py` to run the full gpt-oss model in tinygrad and torch and you can compare their outputs.

**Comment by eitanturok** [comment]
@wozeparrot can i get the bounty locked?

gpt-oss Inference matches torch/hf. Next, I need to focus on making this faster.

**Comment by geohot** [comment]
Oh cool! I'll lock the bounty for this

**Comment by eitanturok** [comment]
On H100, BS=1 fp32 and bf16 both have very slow throughput. But memory is cut in half with bf16 so we'll use it.

fp32
Screenshot 2025-11-16 at 9 17 16 PM
bf16
Screenshot 2025-11-16 at 9 30 03 PM

**Comment by eitanturok** [comment]
The code is slow because
1. 9678us - softmax the expert logits and scatter them to make expert probs. In comparison, computing attention takes 9us.
2. 1049us - for add, matmul, getitem. I think the getitem makes this slow??
3. topk makes a bunch of (recursive?) calls (this is very small impact)

`DEBUG=2 PYTHONPATH=. python tinygrad/apps/llm2.py --timing`
Screenshot 2025-11-16 at 9 43 53 PM

**Comment by eitanturok** [comment]
m4-max olmoe  runs at 104.1 tok/s [source](https://discord.com/channels/1068976834382925865/1068976834928193609/1440081386253062305)

**Comment by eitanturok** [comment]
` PYTHONPATH=. GPT_OSS_LAYERS=1 python tinygrad/apps/llm2.py` generates several slow kernels

```*** METAL     75 r_8_120_4_2_16_3_3_720_4                       arg  6 mem   3.98 GB tm     33.82ms/  1986.66ms (    220 GFLOPS   32|210    GB/s) ['__add__', '__mul__', '__matmul__', '__getitem__']```

```*** METAL    102 r_60_2_16_3_3_2880_32                          arg  6 mem   3.98 GB tm    236.09ms/  2223.06ms (     47 GFLOPS    2|22     GB/s) ['sigmoid', 'clamp', '__matmul__', '__mul__', 'sum', '__add__']```

**Comment by eitanturok** [comment]
1. Before I was incorrectly measured prefill throughput. That's fixed.
2. I got jit to work! When using one-layer, we get 8 tok/s with JIT=2 and 2 tok/s otherwise on my M1 Mac Pro. This will def be slower for the full model but I'm on my local mac and don't have enough room to load the full model. The average throughput is lower than 8 tok/s because we start measuring before the JIT captures.

Screenshot 2025-11-26 at 12 04 46 AM

Screenshot 2025-11-26 at 12 04 57 AM

**Comment by eitanturok** [comment]
[JIT](https://mesozoic-egg.github.io/tinygrad-notes/20240102_jit.html) questions:
1. `JIT=1` or `JIT=2`?
2. Can we remove the `.contiguous()` in RoPE? The comment says this is needed for JIT...
3. How does JIT interact with `topk`? `topk` calls `sort` which makes a bunch of recursive calls. I'm worried that we are making the same recursive calls regardless of the data changing due to the JIT. Why does/doesn't this issue arise?

`VIZ=1 JIT=0 PORT=8000 PYTHONPATH=. GPT_OSS_LAYERS=1 DEBUG=2 python tinygrad/apps/llm2.py --timing --max-new-tokens 5 --prompt I`
* every forward pass schedules 14 kernels, 52 kernels
* last token: 1.86 tok/s

`VIZ=1 JIT=1 PORT=8080 PYTHONPATH=. GPT_OSS_LAYERS=1 DEBUG=2 python tinygrad/apps/llm2.py --timing --max-new-tokens 5 --prompt I`
* every non-JIT-ed forward pass schedules 8 kernels, 46 kernels
* every JIT-ed forward pass batches 32, batched 22 kernels
* last token: 8.30 tok/s

Why does gpt-oss with `JIT=0` have 14, 52 kernels scheduled but `JIT=1` has 8, 46 kernels schedued in non-JIT-ed layers? When `JIT=1`, the non-JIT-ed forward pass use `self.forward_jit` and `start_pos` is a Uop variable. But when `JIT=0`, the forward pass uses `self.forward` and `start_pos` is an int. These differences create different input graphs in the scheduler which results in `schedule_with_vars` and `get_rangeify_map` breaking up the kernels differently. Moreover, we have different numbers of kernels 14+52=66 != 8+46=54.

How does OLMoE hold up?

`VIZ=1 JIT=1 PORT=8002 PYTHONPATH=. OLMOE_LAYERS=1 DEBUG=2 python examples/olmoe.py`
* every non-JIT-ed forward pass schedules 7 kernels, 18 kernels
* every JIT-ed forward pass batches 25 kernels
* last token:  13.40 tok/s

**Comment by eitanturok** [comment]
Why do we have `start_pos` in shape of gpt-oss but now olmoe when jit=1? I think of the attention sink...

`VIZ=1 JIT=0 PORT=8000 PYTHONPATH=. GPT_OSS_LAYERS=1 DEBUG=2 python tinygrad/apps/llm2.py --timing --max-new-tokens 5 --prompt I`

Screenshot 2025-11-27 at 3 17 07 AM


`VIZ=1 JIT=0 PORT=8002 PYTHONPATH=. OLMOE_LAYERS=1 DEBUG=2 python examples/olmoe.py`

Screenshot 2025-11-27 at 3 21 34 AM

**Comment by eitanturok** [comment]
I'm trying to speedup gpt-oss with `JIT=1` on my M1 Mac Pro. and had some questions about the profiler
1. Why does the light blue comile events only appear in the first step? 
2. Why does the metal graph appear in the fourth step (step=3) and not the third step (step=2)? I thought the JIT records the compute graph during the second step (step=1) and resues is for the rest of the steps (step>=2). But for some reason
2. Why does the 1st step take longer than the other steps? We already initialized the weights and casted them to bf16 before starting step 0, right?
3. We record the JIT 
4. What's the difference between JIT=1 and JIT=2?

My best answers so far:
1. The light blue compile events is where we turn the UOp AST into Metal (byte) code. This only appears in the first step because tinygrad caches this. Tinygrad always caches this, even if `JIT=0`.

**Comment by eitanturok** [comment]
This is even slow on an H100... at 3 tok/sec and with beam search gets ~22 tok/sec. What is going on? We get so few FLOPS
Screenshot 2025-12-01 at 1 11 28 AM

`DEBUG=2 VIZ=1 PYTHONPATH=. python tinygrad/apps/llm2.py --timing --prompt I --max-new-tokens 5` 
`DEBUG=2 VIZ=1 PYTHONPATH=. BEAM=2 python tinygrad/apps/llm2.py --timing --prompt I --max-new-tokens 5`
